crates/wrapper/src/lib.rs

```diff
@@ -0,0 +1,133 @@
+pub struct ZombieNode {
+    name: String,
+    // other fields
+}
+
+impl ZombieNode {
+    pub fn new(name: String) -> Self {
+        // name can't be empty!
+        ZombieNode {
+            name, // other fields with default initialization
+        }
+    }
+
+    // consume and return a new object with the modifications
+    // to allow user to chain operations
+    pub fn otherMethod(mut self) -> Self {}
+}
+
+// ------
+
+pub struct ZombieCollator {
+    name: String,
+    // other fields
+}
+
+impl ZombieNode {
+    pub fn new(name: String) -> Self {
+        // name can't be empty!
+        ZombieCollator {
+            name, // other fields with default initialization
+        }
+    }
+
+    // consume and return a new object with the modifications
+    // to allow user to chain operations
+    pub fn otherMethod(mut self) -> Self {}
+}
+
+
+use std::marker::PhantomData;
+
+struct EmptyState;
+struct IdOnly;
+struct CollatorOnly;
+struct IdAndCollator;
```

Comment:
Downstream users can extend state sets and possibly break their guarantees. If you're looking to stop that, take a look into the following:

* https://rust-lang.github.io/api-guidelines/future-proofing.html
* https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/
* https://docs.rs/sealed/latest/sealed/

Reply:
I know about sealed traits and was looking for a good place to utilize them, this will be fun!

> While the typestate pattern makes the builder safe by construction, I believe it also makes the DX worse for both developers and users and as such it should be kept only for very critical parts of the system.

Completely agreed. In our convo with @pepoviola we were praising type-state pattern a lot, and maybe I overdid a bit :D 

I'll tune it down for the DX and try to utilize sealed traits instead. 

These are fantastic suggestions, appreciated 👍

---

I think we all are more or less on the same page here, so I'll start the implementation after revizing this